File: DailyChallenges/Scala/LeetCode_1443.scala
object Solution {
    def minTime(n: Int, edges: Array[Array[Int]], hasApple: List[Boolean]): Int = {
        val m = scala.collection.mutable.HashMap.empty[Int, List[Int]]
        for (e <- edges){
            m(e(0)) = e(1) :: m.getOrElse(e(0), Nil)
            m(e(1)) = e(0) :: m.getOrElse(e(1), Nil)
        }
        val hasApp = hasApple.toArray
        var ans = 0
        def cntTime(node:Int, p: Int): Boolean = {
            var ha = hasApp(node)
            for (ch <- m(node) if ch != p; cha = cntTime(ch, node)){
                if (cha) ans += 1
                ha = ha | cha
            }
            ha
        }
        cntTime(0,-1)
        ans*2
    }
}

File: DailyChallenges/CPP/LeetCode_1443.cpp
class Solution {
public:
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        unordered_map<int, vector<int>> m;
        for (auto& e: edges)
            m[e[0]].push_back(e[1]), m[e[1]].push_back(e[0]);

        int ans = 0;
        function<bool(int, int)> cntTime = [&](int node, int parent) {
            bool ha = hasApple[node];
            for (int ch: m[node]){
                if (ch != parent){
                    bool cha = cntTime(ch, node);
                    if (cha) ++ans;
                    ha = ha | cha;
                }
            }
            return ha;
        };
        cntTime(0,-1);
        return ans*2;
    }
};
